ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1531 @@
+		/* Can't convert from `string->encoding` to `toenc` */
 		return NULL;
 	}
 
-	if (filter2 != NULL) {
-		filter2->illegal_mode = MBFL_OUTPUTFILTER_ILLEGAL_MODE_CHAR;
-		filter2->illegal_substchar = 0x3f;		/* '?' */
-	}
-
 	mbfl_memory_device_init(&device, string->len, (string->len >> 2) + 8);
-
-	/* feed data */
-	n = string->len;
-	p = string->val;
-	if (p != NULL) {
-		while (n > 0) {
-			if ((*filter1->filter_function)(*p++, filter1) < 0) {
-				break;
-			}
-			n--;
-		}
-	}
-
+	mbfl_filter_feed_data(string, filter1);
 	mbfl_convert_filter_flush(filter1);
 	mbfl_convert_filter_delete(filter1);
-	if (filter2 != NULL) {
-		mbfl_convert_filter_flush(filter2);
+	if (filter2) {
 		mbfl_convert_filter_delete(filter2);
 	}
 
+	result->encoding = toenc;
 	return mbfl_memory_device_result(&device, result);
 }
 
+/*
+ *  strlen
+ */
+size_t mbfl_strlen(const mbfl_string *string)
+{
+	return byte_offset_to_char_offset(string->encoding, string->val, string->len, string->len);
+}
 
 /*
- * identify encoding
+ *  strpos
  */
-const mbfl_encoding *
-mbfl_identify_encoding(mbfl_string *string, const mbfl_encoding **elist, int elistsz, int strict)
+size_t mbfl_strpos(mbfl_string *haystack, mbfl_string *needle, ssize_t offset, int reverse)
 {
-	int i, num, bad;
-	size_t n;
-	unsigned char *p;
-	mbfl_identify_filter *flist, *filter;
-	const mbfl_encoding *encoding;
-
-	/* flist is an array of mbfl_identify_filter instances */
-	flist = ecalloc(elistsz, sizeof(mbfl_identify_filter));
-
-	num = 0;
-	if (elist != NULL) {
-		for (i = 0; i < elistsz; i++) {
-			if (!mbfl_identify_filter_init2(&flist[num], elist[i])) {
-				num++;
-			}
-		}
-	}
+	ZEND_ASSERT(haystack->encoding == needle->encoding);
 
-	/* feed data */
-	n = string->len;
-	p = string->val;
-
-	if (p != NULL) {
-		bad = 0;
-		while (n > 0) {
-			for (i = 0; i < num; i++) {
-				filter = &flist[i];
-				if (!filter->flag) {
-					(*filter->filter_function)(*p, filter);
-					if (filter->flag) {
-						bad++;
-					}
-				}
-			}
-			if ((num - 1) <= bad && !strict) {
-				break;
-			}
-			p++;
-			n--;
+	unsigned char *offset_pointer = haystack->val;
+	if (offset != 0) {
+		long str_len = mbfl_strlen(haystack); /* `long` because next comparison must be signed */
+		if (offset > str_len || offset < -str_len) {
+			return MBFL_ERROR_OFFSET;
 		}
+		offset_pointer += char_offset_to_byte_offset(haystack->encoding, haystack->val, haystack->len, offset);
 	}
 
-	/* judge */
-	encoding = NULL;
-
-	for (i = 0; i < num; i++) {
-		filter = &flist[i];
-		if (!filter->flag) {
-			if (strict && filter->status) {
- 				continue;
- 			}
-			encoding = filter->encoding;
-			break;
-		}
+	if (haystack->len < needle->len) {
+		return MBFL_ERROR_NOT_FOUND;
 	}
 
-	/* fall-back judge */
-	if (!encoding) {
-		for (i = 0; i < num; i++) {
-			filter = &flist[i];
-			if (!filter->flag && (!strict || !filter->status)) {
-				encoding = filter->encoding;
-				break;
-			}
-		}
+	const char *found_pos;
```

Comment:
It looks like the conversion to UTF-8 has been dropped from this function. This conversion was there for a reason: We can only implement mbfl_strpos using a simple binary strpos if the encoding is self-synchronizing. Consider a two-byte encoding and a string `AA|BB|AB` in which we search for `AB`. Then as implemented, we will find `A|B` between two characters, which is incorrect.

Now, the conversion to UTF-8 is a bit overzealous in that there are other self-synchronizing encodings (notably all the single byte encodings), but it should still be there for non-self-synchronizing encodings.

It would be great to add some test coverage for this.

Reply:
Thanks for raising this important issue. I would like to look through all the encodings supported by mbstring and determine exactly which ones are affected by this issue.

I would prefer only to convert to/from UTF-8 *if* the input encoding is affected.

And yes: whatever I find will receive corresponding test coverage.

Reply:
Done.